core/indexer/process/transactions/transactionsProcessor.go

```diff
@@ -0,0 +1,239 @@
+package transactions
+
+import (
+	"encoding/hex"
+
+	logger "github.com/ElrondNetwork/elrond-go-logger"
+	"github.com/ElrondNetwork/elrond-go/core"
+	"github.com/ElrondNetwork/elrond-go/core/indexer/types"
+	"github.com/ElrondNetwork/elrond-go/data"
+	"github.com/ElrondNetwork/elrond-go/data/block"
+	"github.com/ElrondNetwork/elrond-go/data/smartContractResult"
+	"github.com/ElrondNetwork/elrond-go/data/transaction"
+	"github.com/ElrondNetwork/elrond-go/process"
+	"github.com/ElrondNetwork/elrond-go/sharding"
+)
+
+const (
+	// A smart contract action (deploy, call, ...) should have minimum 2 smart contract results
+	// exception to this rule are smart contract calls to ESDT contract
+	minimumNumberOfSmartContractResults = 2
+)
+
+var log = logger.GetOrCreate("indexer/process/transactions")
+
+type txDatabaseProcessor struct {
+	txFeeCalculator    process.TransactionFeeCalculator
+	txBuilder          *txDBBuilder
+	txGrouper          *txGrouper
+	txLogsProcessor    process.TransactionLogProcessorDatabase
+	saveTxsLogsEnabled bool
+}
+
+// NewTransactionsProcessor will create a new instance of transactions database processor
+func NewTransactionsProcessor(
+	addressPubkeyConverter core.PubkeyConverter,
+	txFeeCalculator process.TransactionFeeCalculator,
+	isInImportMode bool,
+	shardCoordinator sharding.Coordinator,
+	saveTxsLogsEnabled bool,
+	calculateHash func(object interface{}) ([]byte, error),
+	txLogsProcessor process.TransactionLogProcessorDatabase,
+) *txDatabaseProcessor {
+	txBuilder := newTransactionDBBuilder(addressPubkeyConverter, shardCoordinator, txFeeCalculator)
+	txDBGrouper := newTxGrouper(txBuilder, calculateHash, isInImportMode, shardCoordinator.SelfId())
+
+	if isInImportMode {
+		log.Warn("the node is in import mode! Cross shard transactions and rewards where destination shard is " +
+			"not the current node's shard won't be indexed in Elastic Search")
+	}
+
+	return &txDatabaseProcessor{
+		txFeeCalculator:    txFeeCalculator,
+		txBuilder:          txBuilder,
+		txGrouper:          txDBGrouper,
+		saveTxsLogsEnabled: saveTxsLogsEnabled,
+		txLogsProcessor:    txLogsProcessor,
+	}
+}
+
+// PrepareTransactionsForDatabase will prepare transactions for database
+func (tdp *txDatabaseProcessor) PrepareTransactionsForDatabase(
+	body *block.Body,
+	header data.HeaderHandler,
+	txPool map[string]data.TransactionHandler,
+) ([]*types.Transaction, []*types.ScResult, []*types.Receipt, map[string]*types.AlteredAccount) {
```

Comment:
4 output arguments: please use a DTO that will hold them

Reply:
done.